File: core/mux.py
"""Frame codec + smux-style multiplexer for Zpoint.

Frame JSON: {"i": seq, "s": stream, "t": type, "b": b64(ciphertext)}
types: o=open(host inside), d=data, c=close, k=ping, w=window credit

AUDIT HARDENING (v2):
  * _push() now assigns the transport seq AND hands the frame to send_fn
    under ONE critical section.  The old code released the lock between
    numbering and enqueue, so two threads could interleave seq 41 and 42
    and enqueue them 42-then-41 — downstream reassembly corrupted even
    with an ordered WriteQueue.  Ordering now starts at the mux itself.
  * deliver() uses a per-mux lock for the stream table + credit maps (the
    old dict mutations raced with _drop() from other threads).
  * recv() blocks on a per-stream Condition instead of a 5 ms sleep poll
    (lower latency, no busy wakeups, no lost-wakeup window).
  * close_stream() is idempotent; _drop() wakes all recv() waiters.
  * write() raises ConnectionError after the credit deadline instead of
    silently returning (callers can tear the stream down deterministically).

Property of the @ily_bio research channel (@iliyahsatam).
"""
from __future__ import annotations

import base64
import json
import struct
import threading
import time
import zlib
from collections import deque
from typing import Callable, Optional

T_OPEN, T_DATA, T_CLOSE, T_PING, T_CREDIT = "o", "d", "c", "k", "w"
MAX_PLAIN = 60 * 1024          # 60 KiB plaintext cap per frame
B64 = base64.urlsafe_b64encode


def UNB64(s) -> bytes:
    """Padding-tolerant urlsafe base64 decode.

    BUGFIX #21 (31 Aug): the Kotlin app encodes the INNER mux frame body
    with `Base64.getUrlEncoder().withoutPadding()`, while this module
    called bare `base64.urlsafe_b64decode` — which REQUIRES padding.
    Frames whose unpadded length % 4 != 0 raised binascii.Error
    ("Incorrect padding") on the exit and were silently dropped: the
    stream opened (short OPEN frames often survived), the uplink
    "worked", and the downlink never answered ("upload sent, no
    download").  This is the inner-frame half of the #17 envelope fix.
    """
    if isinstance(s, str):
        s = s.encode()
    return base64.urlsafe_b64decode(s + b"=" * (-len(s) % 4))


def pack_frame(seq: int, stream: int, ftype: str, body: bytes,
               crypto) -> dict:
    """zlib-compress then encrypt a frame body; returns the JSON dict."""
    comp = zlib.compress(body, 6) if body else b""
    ct = crypto.seal(seq, comp, aad=struct.pack("!IB", stream, ord(ftype)))
    return {"i": seq, "s": stream, "t": ftype, "b": B64(ct).decode()}


def unpack_frame(obj: dict, crypto) -> tuple[int, int, str, bytes]:
    seq = int(obj["i"])
    stream = int(obj["s"])
    ftype = str(obj["t"])
    ct = UNB64(obj["b"].encode())
    aad = struct.pack("!IB", stream, ord(ftype))
    comp = crypto.open(seq, ct, aad=aad)
    return seq, stream, ftype, zlib.decompress(comp) if comp else b""


class Mux:
    """smux-lite over an async send() callback.

    Exactly ONE crypto per direction:
      - the *sending* crypto is passed to __init__ (used by pack_frame)
      - the *receiving* crypto is passed to deliver() (used by unpack_frame)

    Flow control: receiver grants T_CREDIT of `window//2` bytes every time
    its cumulative per-stream received bytes cross another half-window
    threshold (tracked in _recv_bytes / _recv_granted).
    """

    def __init__(self, send_fn: Callable[[dict], None], crypto,
                 is_server: bool = False, window: int = 512 * 1024,
                 client_id: str = "c1"):
        self._send = send_fn
        self.crypto = crypto            # for OUTGOING frames (this side)
        self.client_id = client_id      # embedded in OPEN frames
        self._seq = 0
        self._io_lock = threading.Lock()      # seq assignment + send order
        self._next_sid = 0 if is_server else 1
        self._sid_lock = threading.Lock()
        self.window = window
        self._grant = window // 2
        # outbound credit accounting (what the PEER allows us to send)
        self._tx_credits: dict[int, int] = {}
        self._tx_used: dict[int, int] = {}
        # inbound accounting (what WE must credit back)
        self._recv_bytes: dict[int, int] = {}
        self._recv_granted: dict[int, int] = {}
        self._rx_lock = threading.RLock()     # inbound tables + credits
        self._streams: dict[int, deque] = {}
        self._recv_cond = threading.Condition(self._rx_lock)
        self.on_open: Optional[Callable[[int, tuple], None]] = None
        self.on_data: Optional[Callable[[int, bytes], None]] = None
        self.on_close: Optional[Callable[[int], None]] = None
        self.closed = threading.Event()

    # ---------- outbound ----------
    def open_stream(self, target: tuple[str, int]) -> int:
        with self._sid_lock:
            sid = self._next_sid
            self._next_sid += 2
        with self._rx_lock:
            self._streams[sid] = deque()
            self._tx_credits[sid] = self.window
            self._tx_used[sid] = 0
        # OPEN payload: "<client_id>@host:port" so the exit learns where to
        # write downlink frames for this client (BUGFIX #3).
        meta = f"{self.client_id}@{target[0]}:{target[1]}".encode()
        self._push(sid, T_OPEN, meta)
        return sid

    def write(self, sid: int, data: bytes,
              credit_timeout: float = 30.0) -> None:
        for off in range(0, len(data), MAX_PLAIN):
            chunk = data[off:off + MAX_PLAIN]
            self._wait_credit(sid, len(chunk), credit_timeout)
            if self.closed.is_set():
                raise ConnectionError("mux closed")
            self._push(sid, T_DATA, chunk)

    def close_stream(self, sid: int) -> None:
        """Idempotent close: exactly one T_CLOSE per stream per side."""
        first = False
        with self._io_lock:
            # membership check under the same lock that _drop() uses,
            # so double-close can never double-send T_CLOSE.
            if sid in self._streams or sid in self._tx_credits:
                first = True
        if not first:
            return
        if not self.closed.is_set():
            self._push(sid, T_CLOSE, b"")
        self._drop(sid)

    def ping(self) -> None:
        self._push(0, T_PING, b"t")

    # ---------- inbound ----------
    def deliver(self, node: dict, rx_crypto) -> None:
        """Decrypt with rx_crypto (the PEER's tx prefix) and dispatch."""
        seq, sid, ftype, body = unpack_frame(node, rx_crypto)
        if ftype == T_OPEN:
            host, _, port = body.decode().rpartition(":")
            with self._rx_lock:
                self._streams.setdefault(sid, deque())
                self._recv_bytes[sid] = 0
                # BUGFIX #12: ack bookkeeping starts at ZERO.  Starting at
                # _grant meant the first T_CREDIT fired only after 512KB
                # received, while senders can only use floor(W/chunk)*chunk
                # (491,520B with 60KB chunks) — a guaranteed credit
                # deadlock on every bulk transfer.
                self._recv_granted[sid] = 0
                # BUGFIX #13: the RECEIVER of an OPEN (the exit for client
                # streams) must also arm its TRANSMIT credit table here.
                # Without it, the T_CREDIT branch's `if sid in
                # self._tx_credits` guard silently dropped every credit the
                # client granted, stalling the downlink after one window.
                self._tx_credits.setdefault(sid, self.window)
                self._tx_used.setdefault(sid, 0)
            if self.on_open:
                self.on_open(sid, (host, int(port)))
        elif ftype == T_DATA:
            with self._rx_lock:
                q = self._streams.get(sid)
                if q is not None:
                    q.append(body)
                    self._recv_cond.notify_all()
                n = self._recv_bytes.get(sid, 0) + len(body)
                self._recv_bytes[sid] = n
                grants: list[bytes] = []
                if sid in self._recv_granted:
                    while n - self._recv_granted[sid] >= self._grant:
                        self._recv_granted[sid] += self._grant
                        grants.append(str(self._grant).encode())
            for g in grants:   # outside the lock: send may enqueue
                self._push(sid, T_CREDIT, g)
            if self.on_data:
                self.on_data(sid, body)
        elif ftype == T_CLOSE:
            if self.on_close:
                self.on_close(sid)
            self._drop(sid)
        elif ftype == T_CREDIT:
            with self._rx_lock:
                if sid in self._tx_credits:
                    self._tx_credits[sid] += int(body)
                    self._recv_cond.notify_all()
        elif ftype == T_PING:
            pass

    def deliver_received(self, sid: int, body: bytes) -> None:
        """Insert already-decrypted payload into a stream's recv queue and
        run the same cumulative credit accounting deliver() does."""
        grants: list[bytes] = []   # NOTE: used below — do not remove again
        with self._rx_lock:
            q = self._streams.get(sid)
            if q is not None:
                q.append(body)
                self._recv_cond.notify_all()
            n = self._recv_bytes.get(sid, 0) + len(body)
            self._recv_bytes[sid] = n
            if sid not in self._recv_granted:
                # BUGFIX #12 (site 3): ack accounting starts at ZERO —
                # lazy-init at _grant deferred the first T_CREDIT to 2*_grant,
                # deadlocking senders at W - (W mod chunk) bytes.
                self._recv_granted[sid] = 0
            while n - self._recv_granted[sid] >= self._grant:
                self._recv_granted[sid] += self._grant
                grants.append(str(self._grant).encode())
        for g in grants:
            self._push(sid, T_CREDIT, g)

    def _push(self, sid: int, ftype: str, body: bytes) -> None:
        if self.closed.is_set():
            return
        # CRITICAL: seq assignment + send_fn must be atomic with respect to
        # other senders, or RTDB can receive 42 before 41 (out-of-order PUTs
        # even with a per-stream ordered WriteQueue).
        with self._io_lock:
            if self.closed.is_set():
                return
            self._seq += 1
            seq = self._seq
            if ftype == T_DATA:
                self._tx_used[sid] = self._tx_used.get(sid, 0) + len(body)
            frame = pack_frame(seq, sid, ftype, body, self.crypto)
            self._send(frame)

    def _wait_credit(self, sid: int, n: int, timeout: float = 30.0) -> None:
        deadline = time.time() + timeout
        with self._recv_cond:
            while (self._tx_credits.get(sid, self.window)
                   - self._tx_used.get(sid, 0)) < n:
                if self.closed.is_set():
                    return
                left = deadline - time.time()
                if left <= 0:
                    raise ConnectionError(
                        f"credit timeout on stream {sid} ({n}B)")
                self._recv_cond.wait(timeout=min(left, 0.25))

    def _drop(self, sid: int) -> None:
        with self._rx_lock:
            self._streams.pop(sid, None)
            self._tx_credits.pop(sid, None)
            self._tx_used.pop(sid, None)
            self._recv_bytes.pop(sid, None)
            self._recv_granted.pop(sid, None)
            self._recv_cond.notify_all()

    # ---------- control-frame helpers (additive, used by daemons that
    # pre-decrypt frames instead of calling deliver()) ----------

    def apply_credit(self, sid: int, nbytes: int) -> None:
        """Peer granted `nbytes` more transmit credit on this stream."""
        with self._rx_lock:
            if sid in self._tx_credits:
                self._tx_credits[sid] += int(nbytes)
                self._recv_cond.notify_all()

    def peer_closed(self, sid: int) -> None:
        """Peer sent T_CLOSE: drop the stream's tx table so writers fail
        fast; wake everyone blocked in recv()/_wait_credit()."""
        self._drop(sid)

    def peer_opened(self, sid: int) -> None:
        """Defensive: create the recv queue for a stream we didn't open."""
        with self._rx_lock:
            self._streams.setdefault(sid, deque())
            self._recv_bytes.setdefault(sid, 0)
            self._recv_granted.setdefault(sid, 0)   # BUGFIX #12 parity

    def recv(self, sid: int, timeout: float = None) -> Optional[bytes]:
        """Pop the next chunk for this stream; None on timeout/close."""
        deadline = None if timeout is None else time.time() + timeout
        with self._recv_cond:
            while True:
                q = self._streams.get(sid)
                if q is None:
                    return None            # stream dropped/closed
                if q:
                    return q.popleft()
                if deadline is not None and time.time() >= deadline:
                    return None
                if self.closed.is_set():
                    return None
                wait = 0.25 if deadline is None else \
                    min(0.25, max(0.0, deadline - time.time()))
                self._recv_cond.wait(timeout=wait)


def encode_body(obj: dict) -> bytes:
    return json.dumps(obj, separators=(",", ":")).encode()


def decode_body(raw: bytes) -> dict:
    return json.loads(raw.decode())

File: android/app/src/main/java/ir/zpoint/client/engine/Mux.kt
package ir.zpoint.client.engine

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * smux-lite multiplexer — port of core/mux.py v2.
 * seq assignment + send enqueue happen atomically (audit target #1 fix);
 * per-stream credit accounting with cumulative half-window grants.
 */
class Mux(
    private val sendFn: (Frame) -> Unit,
    private val crypto: FrameCrypto,
    isServer: Boolean,
    private val clientLabel: String = "c1",
    val window: Int = 512 * 1024
) {
    private val grant = window / 2
    private val seq = AtomicLong(0)
    private val ioLock = Object()                    // seq+send atomicity
    private val nextSid = AtomicInteger(if (isServer) 0 else 1)

    inner class Stream(val sid: Int) {
        val recvQueue = ArrayDeque<ByteArray>()
        @Volatile var open = true
        var txCredits = window.toLong()
        var txUsed = 0L
        var recvBytes = 0L
        // BUGFIX (parity with core #12): ack accounting starts at ZERO.
        // Starting at `grant` deferred the first T_CREDIT to 2*grant
        // (512KB) while senders can only use window - (window mod chunk)
        // — a guaranteed downlink stall after one window.
        var recvGranted = 0L
    }

    val streams = ConcurrentHashMap<Int, Stream>()
    var onOpen: ((Int, String, Int) -> Unit)? = null
    var onData: ((Int, ByteArray) -> Unit)? = null
    var onClose: ((Int) -> Unit)? = null
    /** BUGFIX #22: control-stream handshake frame arrived. */
    var onHello: ((String) -> Unit)? = null
    @Volatile var closed = false

    fun openStream(host: String, port: Int): Int {
        val sid = nextSid.getAndAdd(2)
        streams[sid] = Stream(sid)
        push(sid, Frames.T_OPEN, "$clientLabel@$host:$port".toByteArray())
        return sid
    }

    fun write(sid: Int, data: ByteArray, creditTimeoutMs: Long = 30_000) {
        var off = 0
        while (off < data.size) {
            val end = minOf(off + Frames.MAX_PLAIN, data.size)
            val chunk = data.copyOfRange(off, end)
            waitCredit(sid, chunk.size.toLong(), creditTimeoutMs)
            if (closed) throw java.io.IOException("mux closed")
            push(sid, Frames.T_DATA, chunk)
            off = end
        }
    }

    fun closeStream(sid: Int) {
        val st = streams[sid] ?: return
        if (!st.open) return
        st.open = false
        if (!closed) push(sid, Frames.T_CLOSE, ByteArray(0))
        drop(sid)
    }

    private fun waitCredit(sid: Int, n: Long, timeoutMs: Long) {
        val deadline = System.currentTimeMillis() + timeoutMs
        while (true) {
            val st = streams[sid]
            if (st == null || closed) return
            synchronized(st) {
                if (st.txCredits - st.txUsed >= n) return
            }
            if (System.currentTimeMillis() > deadline)
                throw java.io.IOException("credit timeout on stream $sid (${n}B)")
            Thread.sleep(5)
        }
    }

    private fun push(sid: Int, type: String, body: ByteArray) {
        if (closed) return
        val frame: Frame
        synchronized(ioLock) {
            if (closed) return
            val s = seq.incrementAndGet()
            if (type == Frames.T_DATA) streams[sid]?.let { st -> st.txUsed += body.size }
            frame = Frames.pack(s, sid, type, body, crypto)
        }
        sendFn(frame)   // inside ioLock: preserves enqueue order == seq order
    }

    /** Deliver a decrypted inbound frame (called by the ingest thread). */
    fun deliver(node: Frame, rxCrypto: FrameCrypto) {
        val (sid, type, body) = Frames.unpack(node, rxCrypto)
        when (type) {
            Frames.T_OPEN -> {
                val meta = String(body)
                val at = meta.indexOf('@')
                val hp = if (at >= 0) meta.substring(at + 1) else meta
                val host = hp.substringBeforeLast(':')
                val port = hp.substringAfterLast(':').toInt()
                val st = streams.getOrPut(sid) { Stream(sid) }
                st.open = true
                onOpen?.invoke(sid, host, port)
            }
            Frames.T_DATA -> {
                val st = streams[sid]
                if (st != null) {
                    synchronized(st) {
                        st.recvQueue.addLast(body)
                        (st as Object).notifyAll()
                    }
                    grantsFor(sid, body.size.toLong()).forEach { g ->
                        push(sid, Frames.T_CREDIT, g.toString().toByteArray())
                    }
                    onData?.invoke(sid, body)
                }
            }
            Frames.T_CLOSE -> {
                onClose?.invoke(sid)
                drop(sid)
            }
            Frames.T_CREDIT -> {
                val st = streams[sid] ?: return
                synchronized(st) {
                    st.txCredits += String(body).toLongOrNull() ?: 0
                    (st as Object).notifyAll()
                }
            }
            // BUGFIX #22: liveness frame. Deliberately does NOT touch the
            // stream table -- sid 0 is not a real stream, and creating an
            // entry for it would leak a Stream object per handshake.
            Frames.T_HELLO -> onHello?.invoke(String(body))
            Frames.T_PING -> {}
        }
    }

    /** Account inbound bytes on a stream and return the T_CREDIT grants
     *  owed to the peer (cumulative, half-window) — core grantsFor parity.
     *  Also used internally by deliver() to push credits back. */
    fun grantsFor(sid: Int, n: Long): List<Long> {
        val st = streams[sid] ?: return emptyList()
        val grants = mutableListOf<Long>()
        synchronized(st) {
            st.recvBytes += n
            while (st.recvBytes - st.recvGranted >= grant) {
                st.recvGranted += grant
                grants += grant.toLong()
            }
        }
        return grants
    }

    private fun drop(sid: Int) {
        val st = streams.remove(sid) ?: return
        synchronized(st) {
            st.open = false
            (st as Object).notifyAll()
        }
    }

    // ---------- control-frame helpers (v3 parity with core/mux.py) ----------

    /** BUGFIX #22: send the handshake JSON on the control stream.
     *  Routed through push() so it is zlib-compressed, AES-256-GCM sealed
     *  and given a real monotonic transport seq, exactly like every other
     *  frame. That is the whole point: an ack can only come back if the
     *  peer actually holds the session key. */
    fun sendHello(json: String) {
        push(Frames.CTRL_SID, Frames.T_HELLO, json.toByteArray())
    }

    /** BUGFIX #24: send a properly SEALED T_CREDIT grant.
     *
     *  GasClientNode used to hand-build `Frame(0, sid, T_CREDIT, "262144")`
     *  and pass it straight to sendUp(), bypassing Frames.pack() entirely.
     *  Two consequences, both fatal:
     *    1. the body went on the wire as PLAINTEXT, so the Python exit ran it
     *       through unpack_frame -> AES-GCM open -> InvalidTag and dropped
     *       every single grant ("[gas-exit] bad frame"). The exit's tx
     *       credits were therefore never replenished, and any stream larger
     *       than one 512KB window froze permanently -- the classic
     *       "page loads 90% then hangs";
     *    2. transport seq 0 collided with the GCM nonce of the first real
     *       frame, which is a key-reuse hazard, not just a bug.
     *  push() fixes both: real seq, real seal. */
    fun sendCredit(sid: Int, grant: Long) {
        push(sid, Frames.T_CREDIT, grant.toString().toByteArray())
    }

    /** Peer granted transmit credit (applied by GasClientNode ingest). */
    fun applyCredit(sid: Int, nbytes: Int) {
        val st = streams[sid] ?: return
        synchronized(st) {
            st.txCredits += nbytes
            (st as Object).notifyAll()
        }
    }

    /** Peer sent T_CLOSE: drop the stream (wake recv()/waitCredit). */
    fun peerClosed(sid: Int) { drop(sid) }

    /** Defensive: create the recv queue for a stream we didn't open. */
    fun peerOpened(sid: Int) {
        streams.getOrPut(sid) { Stream(sid) }
    }

    /** Pop next chunk; null on timeout/close. Blocks on monitor, not polling. */
    fun recv(sid: Int, timeoutMs: Long): ByteArray? {
        val st = streams[sid] ?: return null
        val deadline = if (timeoutMs <= 0) 0L
                       else System.currentTimeMillis() + timeoutMs
        synchronized(st) {
            while (true) {
                st.recvQueue.removeFirstOrNull()?.let { return it }
                if (!st.open || closed) return null
                if (timeoutMs <= 0) return null
                val left = deadline - System.currentTimeMillis()
                if (left <= 0) return null
                (st as Object).wait(minOf(left, 250))
            }
        }
    }
}
